content/bayes_nn/bayes.md

```diff
@@ -0,0 +1,318 @@
+
+    def load_dataset(n=150, n_tst=150):
+        np.random.seed(43)
+        def s(x):
+            g = (x - x_range[0]) / (x_range[1] - x_range[0])
+            return 3 * (0.25 + g**2.)
+        x = (x_range[1] - x_range[0]) * np.random.rand(n) + x_range[0]
+        eps = np.random.randn(n) * s(x)
+        y = (w0 * x * (1. + np.sin(x)) + b0) + eps
+        x = x[..., np.newaxis]
+        x_tst = np.linspace(*x_range, num=n_tst).astype(np.float32)
+        x_tst = x_tst[..., np.newaxis]
+        return y, x, x_tst
+
+    y, x, x_tst = load_dataset()
+    ```
+
+### Probabilistic Linear regression
+=== "tensorflow_probability"
+
+    Let's consider you write your network model in a single `tf.function`.
+
+    ```python linenums="1"
+    import tensorflow as tf
+    import tensorflow_probability as tfp
+    tfd = tfp.distributions
+
+    # Build model.
+    model = tf.keras.Sequential([
+    tf.keras.layers.Dense(1),
+    tfp.layers.DistributionLambda(lambda t: tfd.Normal(loc=t, scale=1)),
+    ])
+
+    # Define the loss:
+    negloglik = lambda y, rv_y: -rv_y.log_prob(y)
+
+    # Do inference.
+    model.compile(optimizer=tf.optimizers.Adam(learning_rate=0.05), loss=negloglik)
+    model.fit(x, y, epochs=500, verbose=False)
+
+    # Make predictions.
+    yhat = model(x_tst)
+    ```
+    
+=== "pyro"
+
+    ```python linenums="1"
+    # coding: utf-8
+
+    from pyro.nn import PyroSample
+
+    # Specify model.
+
+    class BayesianRegression(PyroModule):
+        def __init__(self, in_features, out_features):
+            super().__init__()
+            self.linear = PyroModule[nn.Linear](in_features, out_features)
+            self.linear.weight = PyroSample(dist.Normal(0., 1.).expand([out_features, in_features]).to_event(2))
+            self.linear.bias = PyroSample(dist.Normal(0., 10.).expand([out_features]).to_event(1))
+
+        def forward(self, x, y=None):
+            sigma = pyro.sample("sigma", dist.Uniform(0., 10.))
+            mean = self.linear(x).squeeze(-1)
+            with pyro.plate("data", x.shape[0]):
+                obs = pyro.sample("obs", dist.Normal(mean, sigma), obs=y)
+            return mean
+
+
+
+    # Build model.
+    model = BayesianRegression()
+
+    # Fit model given data.
+    coeffs, linear_response, is_converged, num_iter = tfp.glm.fit(
+    model_matrix=features[:, tf.newaxis],
+    response=tf.cast(labels, dtype=tf.float32),
+    model=model)
+    # ==> coeffs is approximately [1.618] (We're golden!)
+
+    # Do inference.
+    model.compile(optimizer=tf.optimizers.Adam(learning_rate=0.01), loss=negloglik)
+    model.fit(x, y, epochs=1000, verbose=False);
+
+    # Profit.
+    [print(np.squeeze(w.numpy())) for w in model.weights];
+    yhat = model(x_tst)
+    assert isinstance(yhat, tfd.Distribution)
+
+    ```
+
+
+The output of the model:
+
+![Placeholder](../images/bayes_nn/lr.png)
+
+
+
+
+## Variational Autoencoder
+
+The generative models could be build using the bayesian neural network.
```

Comment:
"Generative models can be built using a Bayesian neural network. The variational autoencoder is one popular way to forma generative model."